events/equeue/equeue.c

```diff
@@ -438,11 +438,7 @@ void equeue_dispatch(equeue_t *q, int ms) {
         // check if we were notified to break out of dispatch
         if (q->breaks) {
             equeue_mutex_lock(&q->queuelock);
-            if (q->breaks > 0) {
-                q->breaks--;
-                equeue_mutex_unlock(&q->queuelock);
-                return;
-            }
+            q->breaks = false;
```

Comment:
So, unfortunately this logic is subtlely complex...

``` c
if (q->breaks) {    // <- This condition is outside of mutex so it can't be
                    // trusted. Even if this if is true, we're only _probably_
                    // sure that a break was signalled. The reason for the
                    // extra if statement is so that we can avoid getting a
                    // mutex in the common case (no break events).
                    
    equeue_mutex_lock(&q->queuelock);
    if (q->breaks > 0) { // <- Once we're in the mutex, we confirm that break
                         // was signalled
        q->breaks--;
        equeue_mutex_unlock(&q->queuelock);
        return;          // <- This return actually exits equeue_dispatch. Otherwise
                         // the queue happily continues to run.
    }
    equeue_mutex_unlock(&q->queuelock);
}
```

So two problems:
- You still need the second if statement inside the mutex
- You still need the return statement to actually break out of the dispatch loop